packages/typespec-client-generator-core/src/cache.ts

```diff
@@ -0,0 +1,339 @@
+        crossLanguageDefinitionId: getNamespaceFullName(service),
+        subOperationGroups: [],
+      },
+    ];
+  }
+
+  return [];
+}
+
+/**
+ * Create a TCGC operation group for the given type.
+ * This function will also iterate through the type's namespaces and interfaces to build nested operation groups.
+ *
+ * @param context TCGCContext
+ * @param type
+ * @returns
+ */
+function createOperationGroup(
+  context: TCGCContext,
+  type: Namespace | Interface,
+  groupPathPrefix: string,
+): SdkOperationGroup | undefined {
+  let operationGroup: SdkOperationGroup | undefined;
+  const service =
+    findOperationGroupService(context.program, type, context.emitterName) ?? (type as any);
+  if (!isService(context.program, service)) {
+    reportDiagnostic(context.program, {
+      code: "client-service",
+      format: { name: type.name },
+      target: type,
+    });
+  }
+  if (hasExplicitClientOrOperationGroup(context)) {
+    operationGroup = getScopedDecoratorData(context, operationGroupKey, type);
+    if (operationGroup) {
+      operationGroup.groupPath = `${groupPathPrefix}.${getLibraryName(context, type)}`;
+      operationGroup.service = service;
+      operationGroup.subOperationGroups = [];
+
+      if (type.kind === "Namespace") {
+        operationGroup.subOperationGroups =
+          buildHierarchyOfOperationGroups(context, type, operationGroup.groupPath) ?? [];
+      }
+    }
+  } else {
+    // if there is no explicit client, we will treat non-client namespaces and all interfaces as operation group
+    if (type.kind !== "Interface" || !isTemplateDeclaration(type)) {
+      operationGroup = {
+        kind: "SdkOperationGroup",
+        type,
+        groupPath: `${groupPathPrefix}.${getLibraryName(context, type)}`,
+        service,
+        subOperationGroups: [],
+      };
+    }
+
+    if (operationGroup && type.kind === "Namespace") {
+      operationGroup.subOperationGroups =
+        buildHierarchyOfOperationGroups(context, type, operationGroup.groupPath) ?? [];
+    }
+  }
+
+  // build operation group's cache
+  if (operationGroup) {
+    context.__rawClientsOperationGroupsCache!.set(operationGroup.type!, operationGroup);
+    context.__clientToOperationsCache!.set(operationGroup, []);
+  }
+
+  return operationGroup;
+}
+
+function findOperationGroupService(
+  program: Program,
+  client: Namespace | Interface,
+  scope: LanguageScopes,
+): Namespace | Interface | undefined {
+  let current: Namespace | undefined = client as any;
+  while (current) {
+    if (isService(program, current)) {
+      // we don't check scoped clients here, because we want to find the service for the client
+      return current;
+    }
+    const client = program.stateMap(clientKey).get(current);
+    if (client && (client[scope] || client[AllScopes])) {
+      return (client[scope] ?? client[AllScopes]).service;
+    }
+    current = current.namespace;
+  }
+  return undefined;
+}
+
+function buildHierarchyOfOperationGroups(
```

Comment:
Just checking if my understanding is correct.

We have only 1 level of client / clients. No hierarchy.
Each client can have operation groups, or hierarchy of operation groups (latter only happen on nested Namespace without explicit `@client`)